cirq-core/cirq/transformers/gauge_compiling/multi_moment_gauge_compiling.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2025 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the abstraction for multi-moment gauge compiling as a cirq transformer."""
+
+import abc
+
+import numpy as np
+
+from cirq import circuits, ops
+from cirq.transformers import transformer_api
+
+
+@transformer_api.transformer
+class MultiMomentGaugeTransformer(abc.ABC):
+    """A gauge transformer that wraps target blocks of moments with single-qubit gates.
+
+    In detail, a "gauging moment" of single-qubit gates is inserted before a target block of
+    moments. These gates are then commuted through the block, resulting in a corresponding
+    moment of gates after it.
+
+        q₀: ... ───LG0───╭───────────╮────RG0───...
+                         │           │
+        q₁: ... ───LG1───┤  moments  ├────RG1───...
+                         │   to be   │
+        q₂: ... ───LG2───┤ gauged on ├────RG2───...
+                         │           │
+        q₃: ... ───LG3───╰───────────╯────RG3───...
+    """
+
+    def __init__(
+        self,
+        target: ops.Gate | ops.Gateset | ops.GateFamily,
+        supported_gates: ops.Gateset = ops.Gateset(),
+    ) -> None:
+        """Constructs a MultiMomentGaugeTransformer.
+
+        Args:
+            target: Specifies the two-qubit gates, gate families, or gate sets that will
+              be targeted during gauge compiling. The gauge moment must contain at least
+              one of the target gates.
+            supported_gates: Determines what other gates, in addition to the target gates,
+              are permitted within the gauge moments. If a moment contains a gate not found
+              in either target or supported_gates, it won't be gauged.
+        """
+        self.target = ops.GateFamily(target) if isinstance(target, ops.Gate) else target
```

Comment:
either let mypy know about the existance of these properities by just declaring them

```py3
class MultiMomentGaugeTransformer:
    target: ops.GateFamily
    supported_gates: ops.GateFamily
```

or by making abstract methods that return them